### PR TITLE
Show progress for fetching workspace (cargo-metadata and loadOutDirsFromCheck)

### LINKED ISSUE #3300: No indication of `cargo metadata` being run
When we open a project, we run `cargo metadata`, then do something with its output, but there's no progress being shown before the "workspace loaded" notification. Loading the project is usually fast (a couple of seconds or less), but `cargo metadata` will sometimes want to update the registry index and download the dependencies.

Updating the registry can be especially slow, so we should put up a progress indicator just in case. As for knowing the current status, it prints text, not JSON:

```
warning: please specify `--format-version` flag explicitly to avoid compatibility problems
    Updating crates.io index
  Downloaded fuchsia-zircon v0.3.3
  Downloaded digest v0.8.1
  Downloaded time-macros-impl v0.1.0
[snip]
  Downloaded proc-macro-error v0.4.9
  Downloaded winapi v0.3.8
  Downloaded 168 crates (16.0 MB) in 1.67s (largest was `winapi-x86_64-pc-windows-gnu` at 2.9 MB)
{"packages":[{"name":"thread_local","version":"1.0.1","id":"thread_local 1.0.1 (registry+https://github.com/rust-lang/crates.io-index)","license":"Apache-2.0/MIT","license_file":null,"description":"Per-object thread-local storage","source":"registry+https://github.com/rust-lang/crates.io-index","
```

Related to #3283.

### LINKED ISSUE #7188: No status or progress bar displayed when cargo check for load_out_dirs_from_check
When `load_out_dirs_from_check` is enabled , rust-analyzer will run `cargo check` internally. However, currently we don't display any information to user indicating the checking is running. For a large project (e.g. substrate) it looks like it is hang. 

And I think we face the same problem when RA fetches the meta-data from `cargo-metadata`.. ?

### PR DESCRIPTION
**Pull request by edwin0cheng**
Peek 2021-01-07 21-57

Fixes #7188 
Fixes #3300

### CODE REVIEW

File: crates/stdx/src/lib.rs [hunk 1]
```diff
@@ -147,6 +147,27 @@ where
     left
 }
 
+pub struct JodChild(pub std::process::Child);
+
+impl std::ops::Deref for JodChild {
```
Comment (matklad) [inline crates/stdx/src/lib.rs:152]:
```suggestion
impl ops::Deref for JodChild {
```

File: crates/project_model/src/cargo_workspace.rs [hunk 2]
```diff
@@ -432,6 +443,8 @@ pub(crate) fn load_extern_resources(
                     res.env.insert(package_id, env);
                 }
                 Message::CompilerArtifact(message) => {
+                    reporter.map(|it| it("metadata ".to_string() + &message.target.name));
```
Comment (matklad) [inline crates/project_model/src/cargo_workspace.rs:446]: `format!`

File: crates/project_model/src/workspace.rs [hunk 3]
```diff
@@ -39,6 +39,13 @@ pub enum ProjectWorkspace {
     Json { project: ProjectJson, sysroot: Option<Sysroot> },
 }
 
+#[derive(Debug)]
+pub enum ProjectWorkspaceProgress {
+    Begin,
+    Report(String),
+    End,
+}
+
```
Comment (matklad) [inline crates/project_model/src/workspace.rs:48]: this should be in the `rust-analyzer` crate.

File: crates/project_model/src/workspace.rs [hunk 4]
```diff
@@ -64,7 +71,11 @@ impl fmt::Debug for ProjectWorkspace {
 }
 
 impl ProjectWorkspace {
-    pub fn load(manifest: ProjectManifest, config: &CargoConfig) -> Result<ProjectWorkspace> {
+    pub fn load(
+        manifest: ProjectManifest,
+        config: &CargoConfig,
+        reporter: Option<&Box<dyn Fn(String)>>,
```
Comment (matklad) [inline crates/project_model/src/workspace.rs:77]:
```suggestion
        progress: &mut dyn FnMut(String),
```

**Comment by edwin0cheng** [comment]
Fixed !

**Comment by edwin0cheng** [comment]
bors retry

**Pull request merged**